### PR TITLE
#14 add configuration module with tests

### LINKED ISSUE #14: #### Configuration module compatible with consul.
A configuration module should be capable to:
Read basic configuration values from environment variables: consul credentials and address, http port to listen for requests, postgres schema name, credentials, and client timeout.
Connect to consul and collect information about environment: graylog address, postgres address.

Definition of done:
Configuration is successfully read from environment variables.
Configuration can be read from a consul node.
Unit tests are written and coverage >=80%.

### PR DESCRIPTION
**Pull request by i3odja**
Closes #14

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -2,3 +2,101 @@
...
+	config.consulClient, err = consul.NewClient(consulConfig)
+	if err != nil {
+		return nil, fmt.Errorf("consul client error %w", err)
+	}
+
+	return &config, nil
+}
+
+// Get configurations for glaylog
+func (c *Config) GraylogConfig(ctx context.Context) (string, error) {
```
Comment (oklymchuk) [inline config/config.go]: Suggest replace comment to // GraylogConfig get configurations for glaylog

File: config/config.go [hunk 2]
```diff
@@ -2,3 +2,101 @@
...
+		return "", errors.New("graylog service not found")
+	}
+
+	host := services[0].Address
+	port := services[0].ServicePort
+
+	return fmt.Sprintf("%s:%d", host, port), nil
+}
+
+// Get configuration for Postgres Database
```
Comment (oklymchuk) [inline config/config.go]: Suggest replace comment to // PostgresConfig .....

Comment (crnopster) [reply config/config.go]: or to DBConfig like in PR #16

File: config/config.go (hunk 1 above)
Comment (crnopster) [reply config/config.go]: or to LoggerConfig. (name of package in PR #13 is logger)

File: config/config.go [hunk 3]
```diff
@@ -2,3 +2,101 @@
...
+type Config struct {
+	PostgresUser string `envconfig:"POSTGRES_USER" required:"true"`
+	PostgresPass string `envconfig:"POSTGRES_PASSWORD" required:"true"`
+	PostgresDB   string `envconfig:"POSTGRES_DB" required:"true"`
+
+	ConsulAddress string `envconfig:"CONSUL_ADDRESS" required:"true"`
+	ConsulToken   string `envconfig:"CONSUL_TOKEN" required:"true"`
+
+	BindIP   string `envconfig:"BIND_IP" default:"0.0.0.0"`
+	BindPort int    `envconfig:"BIND_PORT" default:"8000"`
```
Comment (radikh) [inline config/config.go]: What are these values for?

File: cmd/umserver/main.go [hunk 4]
```diff
@@ -27,8 +30,22 @@ func main() {
...
 
+	cfg, err := config.NewConfig()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Example
+	fmt.Println(cfg)
+	fmt.Println(cfg.GraylogConfig(ctx))
+	fmt.Println(cfg.PostgresConfig(ctx))
```
Comment (radikh) [inline cmd/umserver/main.go]: There is already imported log package, please use it.

File: config/config.go (hunk 1 above)
Comment (i3odja) [reply config/config.go]: It is Good advice. Thank you. I fixed it (renamed to LoggerConfig)

File: config/config.go (hunk 2 above)
Comment (i3odja) [reply config/config.go]: Thanks. It is very good idea. I fixed it (renamed to DBConfig)

File: config/config_test.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	}
+	if cfg.PostgresDB != "um_db" {
+		t.Errorf("want um_db got %s", cfg.PostgresDB)
+	}
+	if cfg.ConsulAddress != "consul:8500" {
+		t.Errorf("want consul:8500 got %s", cfg.ConsulAddress)
+	}
+	if cfg.ConsulToken != "token" {
+		t.Errorf("want token got %s", cfg.ConsulToken)
+	}
```
[the remark this answers is not in the document]
Comment (i3odja) [reply config/config_test.go]: Thank you for advice. I changed this test with using assertion library

File: config/config.go (hunk 3 above)
Comment (i3odja) [reply config/config.go]: These values (IP address and Port) are using in ServerAddress() method and return address for http server... 
For example: _Server Listening at 0.0.0.0:8000..._

File: cmd/umserver/main.go (hunk 4 above)
Comment (i3odja) [reply cmd/umserver/main.go]: Thanks, I fixed it

File: cmd/umserver/main.go [hunk 6]
```diff
@@ -27,8 +30,22 @@ func main() {
...
+
 	// TODO: Replace with HTTP server implemented in server package
-	srv := &http.Server{Addr: ":8099"}
+	srv := &http.Server{
+		Addr:         cfg.ServerAddress(),
+		ReadTimeout:  cfg.Timeout,
+		WriteTimeout: cfg.Timeout,
```
[the remark this answers is not in the document]
Comment (i3odja) [reply cmd/umserver/main.go]: Thank you. Done

File: config/config.go (hunk 3 above)
Comment (radikh) [reply config/config.go]: So let them be HTTPIP and HTTPPort, then nobody will have any further questions to what are they.

File: config/config_test.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+	os.Unsetenv("POSTGRES_USER")
+
+	cfg, err = config.NewConfig()
+	require.Error(t, err)
+	assert.Contains(t, err.Error(), "POSTGRES_USER")
+	assert.Nil(t, cfg)
+}
+
+func TestNewConfig_Default(t *testing.T) {
```
Comment (radikh) [inline config/config_test.go]: Use CamelCase for names please.

File: config/config_test.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/lvl484/user-manager/config"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestNewConfig_Required(t *testing.T) {
```
Comment (radikh) [inline config/config_test.go]: Use CamelCase for names please.

File: config/config_test.go (hunk 7 above)
Comment (i3odja) [reply config/config_test.go]: thanks, renamed it

File: config/config_test.go (hunk 8 above)
Comment (i3odja) [reply config/config_test.go]: renamed it too

File: config/config.go (hunk 3 above)
Comment (i3odja) [reply config/config.go]: Good idea. I fixed it

**Pull request merged**